### ISSUE TITLE
Audit: the o2i plugin takes the lattice path only, licensing cannot prefer a dialectal reading, and the case for arbtok.diacritize(text, lect)

### ISSUE THREAD

**Issue by JarbasAl** [TigreGotico/arbtok#98]
> 🤖 Auto-generated by Claude Opus 5 (claude-opus-5) via Claude Code — NOT human-reviewed. Verify before acting.

# Audit: how the o2i plugin uses rawi, and the case for `arbtok.diacritize(text, lect)`

Every claim below is from the code at `dev` or from a run on this machine. Where a figure is quoted, the command that produced it is named.

## 1. The o2i plugin takes the lattice path, always, for every lect

`ArbtokDiacritizer.normalize` (`o2i_plugins.py:77`) constructs `LatticeDiacritizer(lang=lang)` and caches it per language. There is no branch.

**`FusionDiacritizer` is not reachable through the o2i plugin at all.** 452 lines of scoring machinery that the orthography2ipa integration never calls. If fusion is meant to be the path, nothing selects it; if it is not, the plugin docstring should say which one callers get.

## 2. The lattice guards letters, not readings — so rawi's MSA prior wins unopposed by default

The order in `diacritize_word` is: (1) author-complete, (2a) the lect's closed-class lexicon, (2b) the stem lexicon, (2c) proclitic-complete, (3) rawi, then skeleton repair, nisba restoration and (5) licensing.

The lect enters in exactly three places: `underdetermined_positions(..., self._spec)`, `DialectLexicon(lang)`, and `_is_licensed`.

**`_is_licensed` is a veto, not a chooser.** It tokenizes and returns `all(t.kind is not TokenKind.UNKNOWN)` — a boolean over whether every letter and mark is *declared*. It cannot express a preference between two licensed readings.

Measured — six well-formed MSA vocalisations (كِتَاب مَدْرَسَة يَكْتُبُ قَلْب شَجَرَة بَيْت) against four dialect specs:

| spec | licensed |
|---|---|
| ar-EG | 6/6 |
| ar-SA-x-najd | 6/6 |
| ar-MA | 6/6 |
| ar-x-gulf | 6/6 |

**Zero rejections.** A fatḥa is declared in every Arabic spec, so the guard can almost never reject a well-formed MSA vocalisation under a dialect. Where the closed-class lexicon has no entry and the stem lexicon misses, rawi decides alone and the "lattice guards the model" claim does not reach that case. This is the gap a lect-aware `diacritize` would close.

## 3. Licensing draws on the o2i spec's grapheme table, and the گ/چ/ڨ fixes move it

Both paths license through `PhonetokTokenizer(get(lang))` — the spec's own table, not a separate inventory. So a spec gaining or losing a letter changes what is licensed, directly:

| spec | ڨاع | گاعدين | چاي | كتاب |
|---|---|---|---|---|
| ar-DZ | OK | REJECT | REJECT | OK |
| ar-MA | OK | OK | REJECT | OK |
| ar-EG | REJECT | REJECT | OK | OK |
| ar-SA-x-najd | REJECT | OK | OK | OK |

Exactly tracking whether the spec declares the letter. The consequence is operational rather than cosmetic: a word containing an undeclared letter is refused **outright**, so it comes back with **no marks at all** rather than wrong ones.

## 4. The stitching does not pin the model; the call chain does

`EnsembleDiacritizer.__init__` takes `model_path` and `vocab_path`, and rawi v2/v3 share `rawi_v2.vocab.json`, so a re-exported fine-tune that emits pre-argmax logits is loadable in principle — `tools/build_ensemble_logits_onnx.py` reproduces the export.

But `get_ensemble()` is a no-argument `@lru_cache` factory, and `LatticeDiacritizer.diacritizer` calls it with nothing. **There is no plumbing from a diacritizer down to a model path.** Dropping in a register-tuned rawi today needs an edit or a monkeypatch. A `model_path` argument threaded through the diacritizers would be a small change and is worth doing whether or not the rest of this lands.

## 5. Measured on the shipped gold — and the lect layer nearly halves the error

`raw` column in (the gold stripped of ḥarakāt), scored per bare-character position against `sentence`. 881 rows, 44 lects.

| system | rows refused | positions | DER |
|---|---|---|---|
| rawi alone | 67 (7.6%) | 21,292 | **0.1757** |
| lattice | 11 (1.2%) | 22,802 | **0.0989** |
| fusion | 11 (1.2%) | 22,809 | **0.1012** |

Three things this says, in order of how much they should be trusted.

**The lect layer is worth roughly half the diacritic error**, 0.1757 → 0.0989. That is the number the audit was for.

**The gap is understated, not overstated.** A "refusal" here is a skeleton misalignment — the system rewrote letters, so its output cannot be aligned to the reference. rawi alone does that on 7.6% of rows against the lattice's 1.2%, and those rows are *excluded* from rawi's DER. rawi is therefore scored on the 814 rows it did not mangle. Skeleton rewriting is also the worse failure: a wrong mark is a wrong reading, a rewritten letter is a different word.

**Fusion and the lattice are indistinguishable here.** 0.1012 against 0.0989 is about 52 positions in 22,800, roughly one standard error before accounting for within-sentence correlation, which makes the true interval wider still. This gold cannot separate them.

## 6. The gold, its size, and what it can and cannot support

44 lects, 881 sentences, **4,416 words**, 2,943 carrying a mark (66.6%). Median per lect: **20 sentences, 103 words**. Smallest ar-TD at 77 words, largest ar at 120. Files: `data/gold_code_switched/<lect>.tsv`.

**How the marks were produced, which decides whether any of this is circular.** The gold's own docs are explicit and the distinction matters: the `sentence` column is **editor-authored**; the `ipa` column is **pipeline output, never hand-edited** for pinned rows. So scoring a *diacritiser* against `sentence` is **not** circular — unlike `arabic-dialects-gold20`, whose IPA came from correcting our own output. Scoring a *phonemiser* against `ipa` would be.

Two limits the docs state and this issue repeats rather than buries. The gold is **not native-validated** — editor-authored, grounded in each spec's cited dialectology, unchecked by a speaker. And **frames are reused across lects** with per-zone adaptation of the function words.

**Therefore a per-lect DER is not supportable from this gold.** ~103 words per lect, ~68 of them marked, means a five-point difference is about three positions. The pooled figure is sound; a per-lect table would be a ranking of noise, and the reused frames would make it worse — the between-lect differences concentrate in exactly the function words the closed-class lexicon handles, so a per-lect comparison would largely measure lexicon coverage rather than the systems. The per-lect JSON exists if anyone wants it, with that caveat attached.

## The proposal: `arbtok.diacritize(text, lect)` with per-word provenance

Most of it already exists. `LatticeDiacritizer` tracks `dialect_looked_up`, `looked_up`, `repaired` and `rejected` as parallel lists; what is missing is a per-word record and a public entry point.

```python
diacritize(text: str, lect: str) -> DiacritizedText
# .text     — the diacritized string
# .words    — one record per word: surface, output, and
#             provenance ∈ {author, closed-class, stem-lexicon,
#                           proclitic, model, model-repaired, refused}
```

The provenance is the point, and §2 is why: `model` means rawi's MSA prior decided unopposed, because the lect's only remaining contribution was a veto that cannot fire on a well-formed MSA reading. A corpus filtered to words where the lect actually constrained the answer is a different corpus from one filtered on the lect label, and today nothing lets a caller tell them apart.

**Cost.** Latency is per word and dominated by the ONNX session; the lexicon paths return without touching it, so a sentence's cost scales with the words that reach the model. Both models are already loaded by any caller using the plugin.

**What it cannot do.** It cannot supply a dialect vowel rawi never saw. The inventory constrains letters, not readings, so a vocalisation that is wrong for the variety but well-formed for MSA passes every gate — that is precisely §2, and a provenance field naming `model` is an admission of it rather than a fix.

No mapping is proposed here and none should follow from this without its own citation.
